File: packages/react-native-plugin/android/src/main/java/com/otakit/reactnative/OtaKitRuntime.kt
package com.otakit.reactnative

import android.app.Application
import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.os.UserManager
import android.system.Os
import android.system.OsConstants
import com.facebook.react.ReactHost
import com.facebook.react.bridge.JSBundleLoader
import com.facebook.react.bridge.ReactContext
import com.facebook.react.common.annotations.UnstableReactNativeAPI
import com.facebook.react.fabric.ComponentFactory
import com.facebook.react.runtime.ReactHostDelegate
import com.facebook.react.runtime.ReactHostImpl
import com.otakit.core.ArtifactCache
import com.otakit.core.ArtifactInstaller
import com.otakit.core.EventDelivery
import com.otakit.core.LaunchStore
import com.otakit.core.RNManifest
import java.io.File
import java.net.HttpURLConnection
import java.net.URI
import java.nio.file.Files
import java.util.Base64
import java.util.UUID
import java.util.concurrent.Executors
import java.util.concurrent.Future
import org.json.JSONArray
import org.json.JSONObject

/** One coordinator per application; each native module captures one immutable JS generation. */
object OtaKitRuntime {
  private val stateExecutor = Executors.newSingleThreadExecutor()
  private val networkExecutor = Executors.newSingleThreadExecutor()
  private val main = Handler(Looper.getMainLooper())
  private lateinit var preparation: Future<Engine>
  private var host: ReactHost? = null
  @Volatile private var uiIntent = false
  @Volatile private var foreground = false

  /**
   * Call once in Application.onCreate. All configuration and artifact I/O runs off the main thread.
   */
  @Synchronized
  fun prepare(
    application: Application,
    configuration: () -> String,
    embeddedBundle: () -> File,
    caseFolding: () -> ByteArray,
  ) {
    check(!::preparation.isInitialized) { "OtaKit host is already configured" }
    preparation =
      stateExecutor.submit<Engine> {
        check(application.getSystemService(UserManager::class.java).isUserUnlocked) {
          "STORAGE_UNAVAILABLE"
        }
        val config = JSONObject(configuration())
        val receipt = config.getJSONObject("embeddedReceipt")
        check(
          receipt.getString("framework") == "react-native" &&
            receipt.getString("platform") == "android"
        ) {
          "INCOMPATIBLE_ARTIFACT"
        }
        val builtin =
          JSONObject()
            .put("appId", receipt.getString("appId"))
            .put("platform", "android")
            .put("runtimeVersion", receipt.getString("runtimeVersion"))
            .put("contentHash", receipt.getString("embeddedContentHash"))
            .put("version", receipt.getString("version"))
            .put("bundlePath", embeddedBundle().canonicalPath)
            .put("embedded", true)
        val root = File(application.noBackupFilesDir, "OtaKitRN").canonicalFile
        Files.createDirectories(root.toPath())
        val sync = LaunchStore.DirectorySync { directory ->
          val fd = Os.open(directory.path, OsConstants.O_RDONLY or OsConstants.O_NOFOLLOW, 0)
          try {
            check(OsConstants.S_ISDIR(Os.fstat(fd).st_mode)) { "Expected storage directory" }
            Os.fsync(fd)
          } finally {
            Os.close(fd)
          }
        }
        val installer = ArtifactInstaller(caseFolding())
        val cache =
          ArtifactCache(
            root,
            builtin,
            installer,
            keys(config.getJSONObject("publicKeys")),
            config.getString("reactNativeVersion"),
            config.getInt("hermesBytecodeVersion"),
            sync,
          )
        val store =
          LaunchStore(
            File(root, "launch.json"),
            config.getString("nativeBuildId"),
            builtin,
            sync,
            cache::verify,
          )
        // Serialize cleanup with downloads while allowing RN startup to proceed independently.
        networkExecutor.execute { runCatching { cache.collect(store.state()) } }
        if (!config.isNull("ingestURL")) {
          runCatching {
            val info = application.packageManager.getPackageInfo(application.packageName, 0)
            val build =
              if (android.os.Build.VERSION.SDK_INT >= 28) info.longVersionCode.toString()
              else info.versionCode.toString()
            EventDelivery(
                store,
                config.getString("ingestURL"),
                builtin.getString("appId"),
                "android",
                build,
                config.optBoolean("allowLocalhost"),
              )
              .start()
          }
        }
        Engine(config, builtin, root, store, installer, cache)
      }
    main.post(
      object : Runnable {
        override fun run() {
          if (preparation.isDone && !preparation.isCancelled)
            stateExecutor.execute {
              runCatching {
                if (engine().store.checkTimeout(SystemClock.elapsedRealtime(), 10_000)) reload()
              }
            }
          main.postDelayed(this, 250)
        }
      }
    )
  }

  /** Set before creating an Activity's RN surface. Headless starts leave staged code untouched. */
  fun setForegroundUIIntent(value: Boolean) {
    uiIntent = value
  }

  fun setForeground(value: Boolean) {
    foreground = value
    if (::preparation.isInitialized)
      stateExecutor.execute {
        runCatching { engine().store.setForeground(value, SystemClock.elapsedRealtime()) }
      }
  }

  /**
   * Wrap the application's original delegate so packages, bindings and exception handling survive.
   */
  @OptIn(UnstableReactNativeAPI::class)
  @Synchronized
  fun createHost(
    application: Application,
    original: ReactHostDelegate,
    componentFactory: ComponentFactory,
  ): ReactHost {
    check(host == null) { "OtaKit host already exists" }
    val delegate =
      object : ReactHostDelegate by original {
        override val jsBundleLoader: JSBundleLoader
          get() {
            check(Looper.myLooper() != Looper.getMainLooper()) {
              "Bundle selection must run on the RN background executor"
            }
            val engine = engine()
            engine.store.beginLaunch(uiIntent, true, SystemClock.elapsedRealtime())
            engine.store.setForeground(foreground, SystemClock.elapsedRealtime())
            return JSBundleLoader.createFileLoader(
              engine.store.state().getJSONObject("current").getString("bundlePath")
            )
          }
      }
    // This host always runs verified embedded/OTA bundles. Metro remains a separate development
    // host.
    return ReactHostImpl(application, delegate, componentFactory, false, false).also { created ->
      host = created
      created.addReactInstanceEventListener(
        object : com.facebook.react.ReactInstanceEventListener {
          override fun onReactContextInitialized(context: ReactContext) {
            hostReady(context)
          }
        }
      )
    }
  }

  private fun hostReady(context: ReactContext) {
    val module = context.getNativeModule(OtaKitUpdater::class.java) ?: return
    val generation = module.generation
    engine().store.hostReady(generation)
    val previous = context.jsExceptionHandler
    context.jsExceptionHandler =
      com.facebook.react.bridge.JSExceptionHandler { error ->
        stateExecutor.execute { runCatching { if (engine().store.fail(generation)) reload() } }
        if (previous != null) previous.handleException(error) else throw error
      }
  }

  private fun reload() {
    if (Looper.myLooper() == Looper.getMainLooper())
      host?.reload("OtaKit artifact selection changed")
    else main.post { host?.reload("OtaKit artifact selection changed") }
  }

  private fun engine(): Engine {
    check(::preparation.isInitialized) { "OtaKit host is not configured" }
    return preparation.get()
  }

  internal fun captureGeneration(): Long = engine().store.state().getLong("generation")

  internal fun bind(generation: Long): String {
    val store = engine().store
    store.bindBootstrap(generation)
    val artifact = store.state().getJSONObject("current")
    val context = JSONObject().put("generation", generation.toString())
    for (key in
      listOf("appId", "platform", "runtimeVersion", "contentHash", "releaseId", "channel")) context
      .put(key, artifact.opt(key) ?: JSONObject.NULL)
    context
      .put("artifactRoot", JSONObject.NULL)
      .put("expoConfig", JSONObject.NULL)
      .put("expoDomRoot", JSONObject.NULL)
    if (!artifact.optBoolean("embedded")) {
      val root = File(artifact.getString("bundlePath")).parentFile!!
      context.put("artifactRoot", root.toURI().toString().trimEnd('/'))
      val expo = File(root, "expo-config.json")
      if (expo.exists()) context.put("expoConfig", JSONObject(expo.readText()))
      val descriptor = JSONObject(File(root, "otakit-bundle.json").readText())
      context.put(
        "expoDomRoot",
        descriptor.optJSONObject("expo")?.opt("domRoot") ?: JSONObject.NULL,
      )
    }
    return context.toString()
  }

  internal fun state(): String = engine().store.state().toString()

  internal fun ready(generation: Long) = engine().store.notifyReady(generation)

  internal fun guard(generation: Long, active: Boolean) =
    engine().store.setActivationGuard(active, generation)

  internal fun apply(generation: Long): Boolean {
    check(Looper.myLooper() == Looper.getMainLooper()) {
      "Activation must share the headless-task/UI executor"
    }
    check(foreground && uiIntent) { "ACTIVATION_DEFERRED" }
    val next = engine().store.apply(generation, SystemClock.elapsedRealtime())
    if (next != generation) {
      reload()
      return true
    }
    return false
  }

  internal fun submit(network: Boolean = false, operation: () -> Unit) {
    (if (network) networkExecutor else stateExecutor).execute(operation)
  }

  internal fun submitOnMain(operation: () -> Unit) {
    main.post(operation)
  }

  internal fun backgroundWork(generation: Long) {
    engine().store.observeBackgroundWork(generation)
  }

  internal fun check(generation: Long): String = engine().check(generation)

  internal fun download(generation: Long): String = engine().download(generation)

  private fun keys(value: JSONObject): Map<String, ByteArray> =
    value.keys().asSequence().associateWith { Base64.getDecoder().decode(value.getString(it)) }

  private class Engine(
    val config: JSONObject,
    val builtin: JSONObject,
    val root: File,
    val store: LaunchStore,
    val installer: ArtifactInstaller,
    val cache: ArtifactCache,
  ) {
    private var manifest: JSONObject? = null
    private val channel = if (config.isNull("channel")) null else config.getString("channel")
    private val rnVersion = config.getString("reactNativeVersion")
    private val bytecodeVersion = config.getInt("hermesBytecodeVersion")

    fun check(generation: Long): String {
      store.assertInstance(generation)
      val base = config.getString("cdnURL").trimEnd('/')
      val url =
        "$base/manifests/${builtin.getString("appId")}/v3/android/${channel ?: "__base__"}/${builtin.getString("runtimeVersion")}/manifest.json"
      val downloaded = fetch(url, 4 * 1024 * 1024)
      try {
        val verified =
          RNManifest.verify(
            downloaded.readText(),
            builtin.getString("appId"),
            "android",
            builtin.getString("runtimeVersion"),
            channel,
            keys(config.getJSONObject("publicKeys")),
            System.currentTimeMillis() / 1000,
          )
        store.assertInstance(generation)
        manifest = verified
        return verified.toString()
      } finally {
        downloaded.delete()
      }
    }

    fun download(generation: Long): String {
      store.assertInstance(generation)
      val candidate = checkNotNull(manifest) { "Check for an update before downloading" }
      check(
        candidate.getJSONObject("signature").getLong("exp") > System.currentTimeMillis() / 1000
      ) {
        "EXPIRED_MANIFEST"
      }
      val hash = candidate.getString("contentHash")
      val failed = store.state().getJSONArray("failed")
      for (index in 0 until failed.length()) check(failed.getString(index) != hash) {
        "QUARANTINED_ARTIFACT"
      }
      val pointer =
        JSONObject(builtin.toString())
          .put("contentHash", hash)
          .put("version", candidate.getString("version"))
          .put("releaseId", candidate.getString("releaseId"))
          .put("channel", candidate.opt("channel") ?: JSONObject.NULL)
      if (hash == builtin.getString("contentHash")) {
        check(candidate.getString("version") == builtin.getString("version")) {
          "INCOMPATIBLE_ARTIFACT"
        }
        store.stage(pointer, generation)
        return candidate.toString()
      }
      val destination = cache.directory(hash)
      var staging: File? = null
      var completed = false
      try {
        val inventory: JSONArray
        if (destination.exists()) inventory = cache.verifyDirectory(candidate)
        else {
          val temporary = File(root, "staging/${UUID.randomUUID()}")
          staging = temporary
          Files.createDirectories(temporary.toPath())
          if (candidate.getString("strategy") == "zip") {
            val archive = fetch(candidate.getString("url"), candidate.getLong("size"))
            try {
              inventory =
                installer.installZip(
                  archive,
                  temporary,
                  candidate,
                  keys(config.getJSONObject("bundleKeys")),
                  rnVersion,
                  bytecodeVersion,
                )
            } finally {
              archive.delete()
            }
          } else {
            installer.validateDeltaInventory(candidate)
            val files = candidate.getJSONArray("files")
            for (index in 0 until files.length()) {
              val file = files.getJSONObject(index)
              val downloaded = fetch(file.getString("url"), file.getLong("size"))
              try {
                val target = File(temporary, file.getString("path"))
                Files.createDirectories(target.parentFile!!.toPath())
                Files.move(downloaded.toPath(), target.toPath())
              } finally {
                downloaded.delete()
              }
            }
            inventory =
              installer.verifyDirectory(temporary, candidate, rnVersion, bytecodeVersion, null)
          }
        }
        cache.commit(staging, candidate, candidate.toString(), inventory)
        pointer
          .put("embedded", false)
          .put("bundlePath", File(destination, "index.bundle").canonicalPath)
        store.stage(pointer, generation, staging != null)
        completed = true
        return candidate.toString()
      } finally {
        staging?.deleteRecursively()
        if (!completed)
          runCatching { store.recordDownloadFailure(pointer, "RN download or verification failed") }
      }
    }

    private fun fetch(raw: String, limit: Long): File {
      val uri = URI(raw)
      check(
        uri.userInfo == null &&
          uri.fragment == null &&
          (uri.scheme == "https" ||
            (config.optBoolean("allowLocalhost") &&
              uri.scheme == "http" &&
              uri.host in listOf("localhost", "127.0.0.1", "::1")))
      ) {
        "INVALID_DOWNLOAD_URL"
      }
      val connection = uri.toURL().openConnection() as HttpURLConnection
      connection.instanceFollowRedirects = false
      connection.connectTimeout = 30_000
      connection.readTimeout = 30_000
      val directory = File(root, "downloads")
      Files.createDirectories(directory.toPath())
      val file = File.createTempFile("download-", ".tmp", directory)
      try {
        check(connection.responseCode == 200) { "DOWNLOAD_HTTP_${connection.responseCode}" }
        check(connection.contentLengthLong <= limit) { "DOWNLOAD_TOO_LARGE" }
        val deadline = SystemClock.elapsedRealtime() + 120_000
        var written = 0L
        connection.inputStream.use { input ->
          file.outputStream().use { output ->
            val buffer = ByteArray(8192)
            while (true) {
              check(SystemClock.elapsedRealtime() < deadline) { "DOWNLOAD_TIMEOUT" }
              val count = input.read(buffer)
              if (count == -1) break
              check(count <= limit - written) { "DOWNLOAD_TOO_LARGE" }
              output.write(buffer, 0, count)
              written += count
            }
          }
        }
        return file
      } catch (error: Exception) {
        file.delete()
        throw error
      } finally {
        connection.disconnect()
      }
    }
  }
}

File: packages/react-native-plugin/ios/OtaKitRuntime.swift
import Foundation
import OtaKitUpdaterCore
import UIKit

@objc(OtaKitRuntime)
public final class OtaKitRuntime: NSObject {
  @objc public static let shared = OtaKitRuntime()
  private var store: LaunchStore?
  private var transport: RNTransport?
  private var reload: (() -> Void)?
  private var timer: Timer?
  private var observers: [NSObjectProtocol] = []
  private var preparationStarted = false

  /// Call before creating RN roots. The completion preserves the application's factory/window/props.
  public func prepare(
    configuration: Data, embeddedBundle: URL, caseFoldingData: Data, foregroundUIIntent: Bool,
    reload: @escaping () -> Void, completion: @escaping (Error?) -> Void
  ) {
    precondition(Thread.isMainThread)
    guard !preparationStarted else {
      completion(
        NSError(
          domain: "OtaKit", code: 1,
          userInfo: [NSLocalizedDescriptionKey: "OtaKit host is already configured"]))
      return
    }
    preparationStarted = true
    self.reload = reload
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        let config = try JSONDecoder().decode(RNTransport.Configuration.self, from: configuration)
        let receipt = config.embeddedReceipt
        let builtin = Artifact(
          appId: receipt.appId, platform: "ios", runtimeVersion: receipt.runtimeVersion,
          contentHash: receipt.embeddedContentHash, version: receipt.version,
          bundlePath: embeddedBundle.path, embedded: true)
        guard receipt.framework == "react-native" && receipt.platform == "ios" else {
          throw LaunchError.incompatibleArtifact
        }
        var root = try FileManager.default.url(
          for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true
        ).appendingPathComponent("OtaKitRN")
        try FileManager.default.createDirectory(
          at: root, withIntermediateDirectories: true,
          attributes: [.protectionKey: FileProtectionType.completeUntilFirstUserAuthentication])
        var storageValues = URLResourceValues()
        storageValues.isExcludedFromBackup = true
        try root.setResourceValues(storageValues)
        let installer = try ArtifactInstaller(caseFoldingData: caseFoldingData)
        let cache = ArtifactCache(
          root: root, builtin: builtin, installer: installer, keys: config.publicKeys,
          rnVersion: config.reactNativeVersion, bytecodeVersion: config.hermesBytecodeVersion)
        let engine = try LaunchStore(
          file: root.appendingPathComponent("launch.json"), buildId: config.nativeBuildId,
          builtin: builtin,
          validateCachedArtifact: { try cache.verify($0) })
        let transport = RNTransport(
          config: config, builtin: builtin, root: root, store: engine, installer: installer,
          cache: cache)
        // Collection runs on the transport actor, never on the startup completion path.
        Task { await transport.collectCache() }
        if let rawIngest = config.ingestURL, let ingest = URL(string: rawIngest),
          let delivery = try? EventDelivery(
            store: engine, base: ingest, appId: builtin.appId, platform: "ios",
            nativeBuild: Bundle.main.object(forInfoDictionaryKey: "CFBundleVersion") as? String
              ?? "unknown",
            allowLocalhost: config.allowLocalhost == true)
        {
          Task.detached(priority: .utility) {
            while !Task.isCancelled {
              let delay = await delivery.flushOnce()
              if delay > 0 { try? await Task.sleep(nanoseconds: delay * 1_000_000_000) }
            }
          }
        }
        _ = try engine.beginLaunch(
          foreground: foregroundUIIntent, activateStaged: true,
          now: ProcessInfo.processInfo.systemUptime)
        DispatchQueue.main.async {
          self.store = engine
          self.transport = transport
          self.observers = [
            NotificationCenter.default.addObserver(
              forName: UIApplication.didBecomeActiveNotification, object: nil, queue: .main
            ) { _ in engine.setForeground(true, now: ProcessInfo.processInfo.systemUptime) },
            NotificationCenter.default.addObserver(
              forName: UIApplication.willResignActiveNotification, object: nil, queue: .main
            ) { _ in engine.setForeground(false, now: ProcessInfo.processInfo.systemUptime) },
          ]
          self.timer = Timer.scheduledTimer(withTimeInterval: 0.25, repeats: true) { _ in
            do {
              if try engine.checkTimeout(now: ProcessInfo.processInfo.systemUptime) {
                self.reload?()
              }
            } catch {
              // Do not select new code after a failed durable state write.
            }
          }
          completion(nil)
        }
      } catch { DispatchQueue.main.async { completion(error) } }
    }
  }

  @objc public func bundleURL() -> URL? {
    store.map { URL(fileURLWithPath: $0.state().current.bundlePath) }
  }
  @objc public func hostDidStart() {
    if let store = store { store.hostReady(generation: store.state().generation) }
  }
  @objc public func captureGeneration() -> String {
    store.map { String($0.state().generation) } ?? "unconfigured"
  }

  @objc public func bind(_ generation: String) throws -> String {
    let engine = try configuredStore()
    guard let value = UInt64(generation) else { throw LaunchError.staleInstance }
    try engine.bindBootstrap(generation: value)
    let artifact = engine.state().current
    var context: [String: Any] = [
      "generation": generation, "appId": artifact.appId, "platform": artifact.platform,
      "runtimeVersion": artifact.runtimeVersion, "contentHash": artifact.contentHash,
      "releaseId": artifact.releaseId as Any? ?? NSNull(),
      "channel": artifact.channel as Any? ?? NSNull(),
      "artifactRoot": NSNull(), "expoConfig": NSNull(), "expoDomRoot": NSNull(),
    ]
    if !artifact.embedded {
      let root = URL(fileURLWithPath: artifact.bundlePath).deletingLastPathComponent()
      context["artifactRoot"] = root.absoluteString.trimmingCharacters(
        in: CharacterSet(charactersIn: "/"))
      let configFile = root.appendingPathComponent("expo-config.json")
      if let data = try? Data(contentsOf: configFile) {
        context["expoConfig"] = try JSONSerialization.jsonObject(with: data)
      }
      let descriptor =
        try JSONSerialization.jsonObject(
          with: Data(contentsOf: root.appendingPathComponent("otakit-bundle.json")))
        as? [String: Any]
      context["expoDomRoot"] = (descriptor?["expo"] as? [String: Any])?["domRoot"] ?? NSNull()
    }
    return String(data: try JSONSerialization.data(withJSONObject: context), encoding: .utf8)!
  }
  @objc public func stateJSON() throws -> String {
    String(data: try JSONEncoder().encode(configuredStore().state()), encoding: .utf8)!
  }
  @objc public func notifyReady(_ generation: String) throws {
    guard let value = UInt64(generation) else { throw LaunchError.staleInstance }
    try configuredStore().notifyReady(from: value)
  }
  @objc public func setGuard(_ active: Bool, generation: String) {
    guard let value = UInt64(generation) else { return }
    store?.setActivationGuard(active, from: value)
  }
  @objc public func apply(_ generation: String) throws -> NSNumber {
    guard let value = UInt64(generation) else { throw LaunchError.staleInstance }
    let next = try configuredStore().apply(from: value, now: ProcessInfo.processInfo.systemUptime)
    if next != value {
      DispatchQueue.main.async { self.reload?() }
      return true
    }
    return false
  }
  @objc public func check(_ generation: String, completion: @escaping (String?, NSError?) -> Void) {
    guard let transport = transport, let value = UInt64(generation) else {
      completion(nil, NSError(domain: "OtaKit", code: 1))
      return
    }
    Task {
      do { completion(try await transport.check(generation: value), nil) } catch {
        completion(nil, error as NSError)
      }
    }
  }
  @objc public func download(
    _ generation: String, completion: @escaping (String?, NSError?) -> Void
  ) {
    guard let transport = transport, let value = UInt64(generation) else {
      completion(nil, NSError(domain: "OtaKit", code: 1))
      return
    }
    Task {
      do { completion(try await transport.download(generation: value), nil) } catch {
        completion(nil, error as NSError)
      }
    }
  }
  private func configuredStore() throws -> LaunchStore {
    guard let store = store else { throw LaunchError.storageUnavailable }
    return store
  }
}

private actor RNTransport {
  struct Receipt: Decodable {
    let appId: String
    let framework: String
    let platform: String
    let runtimeVersion: String
    let version: String
    let embeddedContentHash: String
  }
  struct Configuration: Decodable {
    let cdnURL: URL
    let channel: String?
    let nativeBuildId: String
    let reactNativeVersion: String
    let hermesBytecodeVersion: UInt32
    let embeddedReceipt: Receipt
    let publicKeys: [String: Data]
    let bundleKeys: [String: Data]
    let allowLocalhost: Bool?
    let ingestURL: String?
  }
  private let config: Configuration
  private let builtin: Artifact
  private let root: URL
  private let store: LaunchStore
  private let installer: ArtifactInstaller
  private let cache: ArtifactCache
  private var manifest: RNManifest?
  private var manifestJSON = "null"
  private var busy = false
  private var collectionPending = true
  init(
    config: Configuration, builtin: Artifact, root: URL, store: LaunchStore,
    installer: ArtifactInstaller, cache: ArtifactCache
  ) {
    self.config = config
    self.builtin = builtin
    self.root = root
    self.store = store
    self.installer = installer
    self.cache = cache
  }
  func check(generation: UInt64) async throws -> String {
    try store.assertInstance(generation)
    guard !busy else { throw LaunchError.activationDeferred }
    busy = true
    defer {
      busy = false
      collectCache()
    }
    let url = config.cdnURL.appendingPathComponent(
      "manifests/\(builtin.appId)/v3/ios/\(config.channel ?? "__base__")/\(builtin.runtimeVersion)/manifest.json"
    )
    let file = try await fetch(url, limit: 4 * 1024 * 1024)
    defer { try? FileManager.default.removeItem(at: file) }
    let data = try Data(contentsOf: file)
    try store.assertInstance(generation)
    manifest = try RNManifest.verify(
      data: data, builtin: builtin, channel: config.channel, keys: config.publicKeys,
      now: Int64(Date().timeIntervalSince1970))
    manifestJSON = String(data: data, encoding: .utf8)!
    return manifestJSON
  }
  func download(generation: UInt64) async throws -> String {
    try store.assertInstance(generation)
    guard !busy, let candidate = manifest else { throw LaunchError.activationDeferred }
    busy = true
    defer {
      busy = false
      collectCache()
    }
    guard candidate.signature.exp > Int64(Date().timeIntervalSince1970) else {
      throw LaunchError.incompatibleArtifact
    }
    guard !store.state().failed.contains(candidate.contentHash) else {
      throw LaunchError.quarantined
    }
    if candidate.contentHash == builtin.contentHash {
      guard candidate.version == builtin.version else { throw LaunchError.incompatibleArtifact }
      try store.stage(
        Artifact(
          appId: builtin.appId, platform: builtin.platform, runtimeVersion: builtin.runtimeVersion,
          contentHash: builtin.contentHash, version: candidate.version,
          releaseId: candidate.releaseId, channel: candidate.channel,
          bundlePath: builtin.bundlePath, embedded: true), from: generation)
      return manifestJSON
    }
    let directory = try cache.directory(candidate.contentHash)
    var completed = false
    defer {
      if !completed {
        try? store.recordDownloadFailure(
          Artifact(
            appId: builtin.appId, platform: builtin.platform,
            runtimeVersion: builtin.runtimeVersion, contentHash: candidate.contentHash,
            version: candidate.version,
            releaseId: candidate.releaseId, channel: candidate.channel,
            bundlePath: directory.appendingPathComponent("index.bundle").path),
          detail: "RN download or verification failed")
      }
    }
    var newStaging: URL?
    let inventory: [RNFileIdentity]
    defer { if let staging = newStaging { try? FileManager.default.removeItem(at: staging) } }
    if FileManager.default.fileExists(atPath: directory.path) {
      inventory = try cache.verifyDirectory(candidate)
    } else {
      let staging = root.appendingPathComponent("staging/\(UUID().uuidString)")
      newStaging = staging
      try FileManager.default.createDirectory(at: staging, withIntermediateDirectories: true)
      if candidate.strategy == "zip" {
        guard let raw = candidate.url, let url = URL(string: raw) else {
          throw LaunchError.incompatibleArtifact
        }
        let archive = try await fetch(url, limit: candidate.size)
        defer { try? FileManager.default.removeItem(at: archive) }
        inventory = try installer.installZip(
          archiveURL: archive, into: staging, manifest: candidate, bundleKeys: config.bundleKeys,
          rnVersion: config.reactNativeVersion, bytecodeVersion: config.hermesBytecodeVersion)
      } else {
        try installer.validateDeltaInventory(candidate)
        let files = candidate.files ?? []
        for file in files {
          guard file.size >= 0 && file.size <= 100 * 1024 * 1024, let url = URL(string: file.url)
          else { throw LaunchError.incompatibleArtifact }
          let downloaded = try await fetch(url, limit: file.size)
          defer { try? FileManager.default.removeItem(at: downloaded) }
          let destination = staging.appendingPathComponent(file.path)
          try FileManager.default.createDirectory(
            at: destination.deletingLastPathComponent(), withIntermediateDirectories: true)
          try FileManager.default.moveItem(at: downloaded, to: destination)
        }
        inventory = try installer.verifyDirectory(
          staging, manifest: candidate, rnVersion: config.reactNativeVersion,
          bytecodeVersion: config.hermesBytecodeVersion)
      }
    }
    try cache.commit(
      staging: newStaging, manifest: candidate, receipt: Data(manifestJSON.utf8), files: inventory)
    try store.stage(
      Artifact(
        appId: builtin.appId, platform: builtin.platform, runtimeVersion: builtin.runtimeVersion,
        contentHash: candidate.contentHash, version: candidate.version,
        releaseId: candidate.releaseId, channel: candidate.channel,
        bundlePath: directory.resolvingSymlinksInPath().appendingPathComponent("index.bundle").path),
      from: generation, downloaded: newStaging != nil)
    completed = true
    return manifestJSON
  }
  private func fetch(_ url: URL, limit: Int64) async throws -> URL {
    try await BoundedDownload.fetch(
      url, limit: limit, allowLocalhost: config.allowLocalhost == true)
  }

  func collectCache() {
    guard collectionPending && !busy else { return }
    collectionPending = false
    // Cleanup failure must not affect the selected code; retry on a later cold start.
    try? cache.collect(store.state())
  }
}
